src/lib.rs

```diff
@@ -1,5 +1,262 @@
+            file_extension: &self.file_extension,
+            message_contains: &self.message_contains,
+            message_equals: &self.message_equals,
+            message_starts_with: &self.message_starts_with,
+        }
+    }
+}
+
+/// The possible input methods.
+#[derive(Subcommand, Debug)]
+pub enum InputMethod {
+    /// Read the input from the local Git history.
+    GitHistory,
+
+    /// Read the specified input file.
+    LogFile {
+        /// The log file to read from.
+        log_file: PathBuf,
+    },
+
+    /// Read from `stdin`.
+    Stdin,
+}
+
+impl InputMethod {
+    /// Process the configured input method.
+    pub fn read(&self) -> Result<String, Box<dyn std::error::Error>> {
+        match self {
+            Self::GitHistory => {
+                let process = std::process::Command::new("git")
+                    .arg("log")
+                    .arg("--numstat")
+                    .output()?;
+
+                Ok(String::from_utf8(process.stdout)?)
+            }
+            Self::LogFile { log_file } => Ok(std::fs::read_to_string(log_file)?),
+            Self::Stdin => {
+                let mut input = String::new();
+
+                loop {
+                    let mut buffer = String::new();
+
+                    if std::io::stdin().read_line(&mut buffer)? == 0 {
+                        break;
+                    }
+
+                    input.push_str(&buffer);
+                }
+
+                Ok(input)
+            }
+        }
+    }
+}
+
+/// The revealed filter creteria.
+///
+/// This data structure allows to filter the input commits by certain creteria.
+#[derive(Debug, Default)]
+pub struct Filter<'a> {
+    /// A set of substrings to be contained by some authors' names.
+    author_contains: &'a [String],
+
+    /// A set of strings to match some authors's names.
+    author_equals: &'a [String],
+
+    /// A set of substrings to be contained by some commits' hashes.
+    commit_contains: &'a [String],
+
+    /// A set of strings to match some commits' hashes.
+    commit_equals: &'a [String],
+
+    /// A set of substrings to be contained by some authors' email
+    /// addresses.
+    email_contains: &'a [String],
+
+    /// A set of strings to match some authors' email addresses.
+    email_equals: &'a [String],
+
+    /// A set of file extensions to filter by.
+    file_extension: &'a [String],
+
+    /// A set of substrings to be contained by some commits' messages.
+    message_contains: &'a [String],
+
+    /// A set of strings to match some commits' messages.
+    message_equals: &'a [String],
+
+    /// A set of strings to introduce some commits' messages.
+    message_starts_with: &'a [String],
+}
+
+impl Filter<'_> {
+    /// Whether the author's email address matches the expectations.
+    fn check_author_email(&self, email: &str) -> bool {
+        let contains =
+            self.email_contains.is_empty() || self.email_contains.iter().any(|e| email.contains(e));
```

Comment:
I try to implement my `cargo-optimise` stricter regarding the Clippy lint levels, at the moment.  When running it over your code, it complained about obsolete closures.  Check what would happen with:

```rust
let contains = self.email_contains.is_empty() || self.email_contains.iter().any(email.contains);
```

Reply:
```shell
error[E0615]: attempted to take value of method `contains` on type `&str`
   --> src\lib.rs:199:84
    |
199 |             self.email_contains.is_empty() || self.email_contains.iter().any(email.contains);
    |                                                                                    ^^^^^^^^ method, not a field
    |
help: use parentheses to call the method
    |
199 |             self.email_contains.is_empty() || self.email_contains.iter().any(email.contains(_));
    |                                                                                            +++

For more information about this error, try `rustc --explain E0615`.
```

does not compile

Reply:
I am sorry, I mixed up the functions.  The redundant closure is in line 104 before your changes, now line 361 after them:

```rust
.map(|l| l.loc())
```

Clippy also offers you some "help: replace the closure with the method itself: `LocDiff::loc`."  But you only get this with `#![deny(clippy::pedantic)]`.

Reply:
I'm not sure, that this is really a better style, since you normally do not call member methods in that way `LocDiff::loc(l)`. I guess, that is why this lint is hidden behind `pedantic`.